core/src/main/java/io/questdb/cairo/wal/WalEventReader.java

```diff
@@ -76,22 +83,72 @@ public WalEventCursor of(Path path, long segmentTxn) {
             );
 
             if (segmentTxn > -1) {
-                // Read record offset and size
-                long fdi = openRO(ff, path.trimTo(pathLen).concat(EVENT_INDEX_FILE_NAME).$(), LOG);
+                final int maxTxn = eventMem.getInt(WALE_MAX_TXN_OFFSET_32);
+                if (maxTxn < -1) {
+                    final int errno = 0;
+                    throw CairoException.critical(errno).put("segment ")
+                            .put(path).put(" does not have a valid maxTxn: ")
+                            .put(maxTxn);
+                }
+
+                // To read the `_event.i` fully we need to map it so we can read the `maxTxn` transaction.
+                // N.B. `+ 2 is for: (+1) to skip to one past the end and (+1) for the size of the data to read.
+                final long fullEventIndexMapSize = (maxTxn + 2L) * Long.BYTES;
+
+                // But if that mapping fails (say, because didn't persist all the data), we want to at least
+                // map until the specific `segmentTxn` we want to read.
+                // N.B. `+ 2` same as above.
+                final long minEventIndexMapSize = (segmentTxn + 2L) * Long.BYTES;
+
+                try {
+                    eventIndexMem.of(
+                            ff,
+                            path.trimTo(pathLen).concat(EVENT_INDEX_FILE_NAME).$(),
+                            ff.getPageSize(),
+                            fullEventIndexMapSize,
+                            MemoryTag.MMAP_TABLE_WAL_READER,
+                            CairoConfiguration.O_NONE,
+                            Files.POSIX_MADV_RANDOM
+                    );
+                } catch (CairoException _couldNotMapToMaxTxn) {
+                    eventIndexMem.of(
```

Comment:
I'm not sure that we get an exception when trying to map beyond the file size on all the platforms. Did you test it on Linux, Windows, and Mac? You can test it in isolation, e.g., create a file and map beyond the created size, and see if you have an exception on map or read.

Reply:
Added test. Executed on Windows.